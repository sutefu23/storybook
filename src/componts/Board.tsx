import { Square } from './Square'
import { StatusBar } from './StatusBar'
import type { Player, Status } from './Game'

export type Props = {
  xIsNext: boolean
  squares: Player[]
  onPlay: (squares: Player[]) => void
  rowCount?: number
  colCount?: number
}
export const Board = ({ xIsNext, squares, onPlay, rowCount = 3, colCount = 3 }: Props) => {
  const winner = calculateWinner(squares)
  let status: Status
  if (winner) {
    status = `Winner: ${winner}`
  } else {
    const isFilledSquare = squares.filter((s) => !!s).length == rowCount * colCount //ゲーム終了判定
    if (isFilledSquare) {
      status = 'Draw!'
    } else {
      status = `Next Player: ${xIsNext ? 'X' : 'O'}`
    }
  }

  const handleClick = (i: number) => {
    if (squares[i] || calculateWinner(squares)) {
      return
    }

    const nextSquares = squares.slice()
    if (xIsNext) {
      nextSquares[i] = 'X'
    } else {
      nextSquares[i] = 'O'
    }
    onPlay(nextSquares)
  }

  return (
    <>
      <StatusBar status={status} />
      {[...new Array(rowCount)].map((_, i) => (
        <div key={i} className="board-row">
          {[...new Array(colCount)].map((_, j) => {
            const index = colCount * i + j
            return <Square key={j} value={squares[index]} onSquareClick={() => handleClick(index)} />
          })}
        </div>
      ))}
    </>
  )
}
function calculateWinner(squares: Player[]) {
  const lines = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
    [0, 3, 6],
    [1, 4, 7],
    [2, 5, 8],
    [0, 4, 8],
    [2, 4, 6]
  ]
  for (let i = 0; i < lines.length; i++) {
    const [a, b, c] = lines[i]
    if (squares[a] && squares[a] === squares[b] && squares[a] === squares[c]) {
      return squares[a]
    }
  }
  return null
}
