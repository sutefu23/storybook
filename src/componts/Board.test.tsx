import { render } from '@testing-library/react'
import * as stories from './Board.stories'
import { composeStories } from '@storybook/testing-react'
// https://lightbulbcat.hatenablog.com/entry/2019/08/08/040659
// https://zenn.dev/takepepe/articles/reusable-story-playfn
const { Default: BoarDefault, FilledTriangle: BoardTriangle } = composeStories(stories)

describe('Boardのスナップショットテスト', () => {
  test('Boardのデフォルト状態のSnapShot', () => {
    const component = render(<BoarDefault />)
    expect(component).toMatchSnapshot()
  })

  test('Boardが△になった状態のSnapShot', () => {
    const component = render(<BoardTriangle />)
    expect(component).toMatchSnapshot()
  })
})
