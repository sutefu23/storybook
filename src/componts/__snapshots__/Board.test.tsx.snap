// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Boardのスナップショットテスト Boardが△になった状態のSnapShot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="status"
      >
        Winner: △
      </div>
      <div
        class="board-row"
      >
        <button
          class="square"
        >
          △
        </button>
        <button
          class="square"
        >
          △
        </button>
        <button
          class="square"
        >
          △
        </button>
      </div>
      <div
        class="board-row"
      >
        <button
          class="square"
        >
          △
        </button>
        <button
          class="square"
        >
          △
        </button>
        <button
          class="square"
        >
          △
        </button>
      </div>
      <div
        class="board-row"
      >
        <button
          class="square"
        >
          △
        </button>
        <button
          class="square"
        >
          △
        </button>
        <button
          class="square"
        >
          △
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="status"
    >
      Winner: △
    </div>
    <div
      class="board-row"
    >
      <button
        class="square"
      >
        △
      </button>
      <button
        class="square"
      >
        △
      </button>
      <button
        class="square"
      >
        △
      </button>
    </div>
    <div
      class="board-row"
    >
      <button
        class="square"
      >
        △
      </button>
      <button
        class="square"
      >
        △
      </button>
      <button
        class="square"
      >
        △
      </button>
    </div>
    <div
      class="board-row"
    >
      <button
        class="square"
      >
        △
      </button>
      <button
        class="square"
      >
        △
      </button>
      <button
        class="square"
      >
        △
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Boardのスナップショットテスト Boardのデフォルト状態のSnapShot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="status"
      >
        次のプレイヤー: O
      </div>
      <div
        class="board-row"
      >
        <button
          class="square"
        />
        <button
          class="square"
        />
        <button
          class="square"
        />
      </div>
      <div
        class="board-row"
      >
        <button
          class="square"
        />
        <button
          class="square"
        />
        <button
          class="square"
        />
      </div>
      <div
        class="board-row"
      >
        <button
          class="square"
        />
        <button
          class="square"
        />
        <button
          class="square"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="status"
    >
      次のプレイヤー: O
    </div>
    <div
      class="board-row"
    >
      <button
        class="square"
      />
      <button
        class="square"
      />
      <button
        class="square"
      />
    </div>
    <div
      class="board-row"
    >
      <button
        class="square"
      />
      <button
        class="square"
      />
      <button
        class="square"
      />
    </div>
    <div
      class="board-row"
    >
      <button
        class="square"
      />
      <button
        class="square"
      />
      <button
        class="square"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
